### PR TITLE
[senku] DrivAerML Random Fourier Features Positional Encoding

### PR DESCRIPTION
**Pull request by morganmcg1**
# [senku] DrivAerML Random Fourier Features Positional Encoding

## Hypothesis

The current `ContinuousSincosEmbed` uses a fixed sinusoidal basis with wavelengths from 1 to max_wavelength. Random Fourier Features (RFF, Rahimi & Recht 2007) sample frequencies from a Gaussian distribution, providing a richer and more adaptive basis for approximating shift-invariant kernels. In the context of CFD point clouds, RFF can capture a broader range of spatial scales simultaneously — both the fine-grained surface curvature (high frequencies) and coarse aerodynamic features (low frequencies) — without manually tuning the frequency schedule.

**Mathematical basis:** RFF approximates the RBF kernel: `exp(-||x-y||²/2σ²) ≈ z(x)^T z(y)` where `z(x) = sqrt(2/D) * [cos(w_1^T x + b_1), ..., cos(w_D^T x + b_D)]` with `w_i ~ N(0, 1/σ²)`.

**Prediction:** RFF with bandwidth `σ=1.0` will outperform sincos PE by learning richer spatial relationships between non-uniformly sampled CFD mesh points, where the standard grid assumptions of sincos PE break down.

## Implementation Instructions

Replace `ContinuousSincosEmbed` with `RFFEmbed` in `model.py`:

```python
class RFFEmbed(nn.Module):
    """
    Random Fourier Features positional encoding.
    Approximates RBF kernel with random frequency sampling.
    Reference: Rahimi & Recht (2007) - Random Features for Large-Scale Kernel Machines
    """
    def __init__(self, hidden_dim: int, input_dim: int = 3, num_freqs: int = 128, sigma: float = 1.0):
        super().__init__()
        # Sample random frequencies from N(0, 1/sigma^2)
        W = torch.randn(input_dim, num_freqs) / sigma
        b = torch.rand(num_freqs) * 2 * math.pi
        self.register_buffer("W", W)
        self.register_buffer("b", b)
        raw_dim = num_freqs * 2  # cos and sin components
        self.proj = nn.Linear(raw_dim, hidden_dim)
    
    def forward(self, coords: torch.Tensor) -> torch.Tensor:
        # coords: [B, N, input_dim]
        coords = coords.float()
        # [B, N, num_freqs]
        angles = coords @ self.W + self.b  # broadcast b
        emb = torch.cat([torch.cos(angles), torch.sin(angles)], dim=-1)  # [B, N, 2*num_freqs]
        return self.proj(emb)
```

**Integration in `model.py`:**
1. Add `import math` at the top if not present
2. Add the `RFFEmbed` class (paste above)
3. In `SurfaceTransolver.__init__`, replace:
   ```python
   self.pos_embed = ContinuousSincosEmbed(...)
   ```
   with:
   ```python
   self.pos_embed = RFFEmbed(hidden_dim=n_hidden, input_dim=space_dim, num_freqs=128, sigma=1.0)
   ```

**Key difference from FourierEmbed:** RFF uses randomly sampled (non-geometric) frequencies that are fixed after initialization, and uses both cos and sin with random phase offsets. This gives a richer, more isotropic coverage of frequency space compared to the geometric progression in FourierEmbed.

**Note:** The random frequencies are sampled at `__init__` time and stored as buffers (not parameters), so they are fixed during training. This is intentional — RFF theory requires the frequencies to be sampled from the kernel distribution, not learned.

## Training Command

```bash
torchrun --nproc_per_node=4 train.py \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --no-use-ema --lr 3e-4 --weight-decay 1e-4 --batch-size 2 \
  --epochs 50 --lr-cosine-t-max 30 \
  --train-surface-points 40000 --train-volume-points 40000 \
  --kill-thresholds "500:train/loss<5,2000:val_primary/abupt_axis_mean_rel_l2_pct<30" \
  --wandb-group bengio-stream2-senku --agent senku
```

**Bonus trial (different bandwidth):** If the first run succeeds, run a second with `sigma=0.5` to compare. Lower sigma = higher frequencies sampled more frequently, which may help with fine-grained surface features.

```bash
torchrun --nproc_per_node=4 train.py \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --no-use-ema --lr 3e-4 --weight-decay 1e-4 --batch-size 2 \
  --epochs 50 --lr-cosine-t-max 30 \
  --train-surface-points 40000 --train-volume-points 40000 \
  --kill-thresholds "500:train/loss<5,2000:val_primary/abupt_axis_mean_rel_l2_pct<30" \
  --wandb-group bengio-stream2-senku --agent senku
```
(change sigma=1.0 to sigma=0.5 in the `RFFEmbed` constructor for the bonus trial)

## Baseline Targets to Beat (AB-UPT Reference)

| Metric | AB-UPT Target |
|--------|--------------|
| `test_primary/surface_pressure_rel_l2_pct` | 3.82 |
| `test_primary/wall_shear_rel_l2_pct` | 7.29 |
| `test_primary/volume_pressure_rel_l2_pct` | 6.08 |
| `test_primary/wall_shear_x_rel_l2_pct` | 5.35 |
| `test_primary/wall_shear_y_rel_l2_pct` | 3.65 |
| `test_primary/wall_shear_z_rel_l2_pct` | 3.63 |
| `test_primary/abupt_axis_mean_rel_l2_pct` | ~4.51 |

**Primary checkpoint metric:** `val_primary/abupt_axis_mean_rel_l2_pct` (lower is better)

## Success Criteria

- Report final `val_primary/abupt_axis_mean_rel_l2_pct` for sigma=1.0 (and sigma=0.5 if run)
- Report all `test_primary/*` metrics
- Compare against FourierEmbed baseline (alphonse run) — does random sampling beat geometric progression?
- Report parameter count of `RFFEmbed` vs `ContinuousSincosEmbed` (should be similar)

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: **Important: Preemptive torch.compile warning** — Multiple students on this wave have reported a crash when `compile_model=True` (the default) hits validation for the first time. The crash is a known PyTorch 2.x Inductor bug in `tiling_utils.get_pw_red_splits` that fires when eval batches have different point-count shapes than training batches.

**If your experiment has not yet run its first validation pass:** Add `--no-compile-model` to your launch command immediately before you hit this bug. Throughput drops from ~20 it/s to ~10-11 it/s but training is stable.

**If your experiment is already past epoch 1 and has not crashed:** You may have gotten lucky with batch shape alignment. Monitor closely at the next validation pass, but you can continue.

No code changes needed — just add `--no-compile-model` to your train.py launch flags.

**Comment by morganmcg1** [comment]
STUDENT senku:

## Status update / heads-up

Previous run crashed with the known torch.compile bug at first validation (PyTorch Inductor tiling_utils AssertionError, exactly as warned in the advisor's preemptive note). Relaunching now with --no-compile-model.

## Research notes (from researcher-agent pass)

Key finding that shaped my plan: DrivAerML coordinates are passed in raw meters (cars span ~5m, volume domain extends ~20m+ into the wake). With W ~ N(0, 1/sigma^2) and sigma=1.0, typical angle std is ||x|| / sigma ≈ 2-20 radians — already a high-frequency regime. Going to sigma=0.5 doubles the frequency further into the same octave, while Tancik et al. 2020 ("Fourier Features Let Networks Learn High Frequency Functions") show the meaningful sigma-sweep range spans multiple octaves.

## Plan / deviation from PR

- **Run 1 (as PR specifies):** sigma=1.0, no compile.
- **Run 2 (deviation, justified):** sigma=10.0 instead of sigma=0.5. This tests the *opposite* octave (lower per-meter frequency content, suited to coarse aerodynamic features) so the two runs together probe whether RFF beats sincos PE under either bandwidth choice. sigma=0.5 is in the same octave as sigma=1.0 and would give little new signal.

If sigma=1.0 is clearly broken and sigma=10.0 unequivocally better (or vice versa), this is much more informative than two adjacent points. Happy to also run sigma=0.5 if you want full coverage.

**Comment by morganmcg1** [comment]
ADVISOR: Approved on both counts.

1. **sigma=10.0 instead of sigma=0.5 for Trial B is the right call.** Your analysis is sound — DrivAerML coordinates are in raw meters, so sigma=1.0 is already deep in the high-frequency regime. Sampling another point in the same octave (sigma=0.5) carries little information. Probing the opposite octave (sigma=10.0) is the more informative experiment and gives us a real bandwidth contrast.

2. **Skip sigma=0.5.** Two well-separated octaves (1.0 vs 10.0) tell us more about the bandwidth landscape than three adjacent points. If either run shows promise, we can do a focused sweep around the winner in a follow-up.

3. **Per the program guidance, please ensure your reported metrics include test evaluated from the best validation checkpoint** (not just terminal-epoch val) — that's our paper-facing comparison number against AB-UPT.

Proceed with Run 1 (sigma=1.0) and Run 2 (sigma=10.0). Looking forward to the comparison.

**Comment by morganmcg1** [comment]
ADVISOR: Last update was 13:37Z (sigma=10.0 approval). Please post current status: which trial is active, current epoch, and val_primary/abupt trajectory. Specifically interested in whether wall_shear metrics differ between sigma=1.0 vs sigma=10.0.

**Comment by morganmcg1** [comment]
ADVISOR: **Check-in — status update required.**

Your W&B runs (2j2j9mf7, 2kk89793, k48trs2y) are all actively training but are showing no validation metrics yet in the W&B summaries. All three appear to be at sigma=1 based on their configs — please clarify.

**Required immediately:**
1. Confirm the sigma values for each of your three runs — are runs 2j2j9mf7, 2kk89793, and k48trs2y testing different sigma values (e.g., sigma=0.1, 1.0, 10.0) or are they all at sigma=1?
2. Are validation metrics being logged? Check that your W&B logging includes `val_primary/abupt_axis_mean_rel_l2_pct` keys.
3. Current step/epoch for each run and ETA to first validation checkpoint
4. Confirm `--no-compile-model` is active for all three runs

If all three runs have sigma=1, this suggests a configuration error — the whole point of your trial matrix was to sweep sigma values. Please check and report back immediately so we can correct if needed.

**Comment by morganmcg1** [comment]
ADVISOR: URGENT — escalation #3, critical bugs in your run set.

Two issues confirmed from W&B inspection:

**1. Sigma sweep never executed.** All three of your active runs (`2j2j9mf7`, `2kk89793`, `k48trs2y`) report `sigma=1.0` in their config, named `senku/rff-sigma1.0-rank*`. The sigma=10.0 trial we agreed on is missing. Did the second job get launched?

**2. Zero validation metrics logged in 25+ epochs of training.** I dumped the full summary keys for `2j2j9mf7` (group `bengio-stream2-senku`, currently at step ~448,820). There are NO `val_primary/*` keys — only train/* and slope/* over training loss. Either:
   - validation is disabled in this run, or
   - the val loop is silently failing, or
   - the run was launched without `--val-every-n-epochs` set correctly.

This run has consumed nearly an entire 50-epoch budget producing zero physically meaningful evaluations. We literally cannot tell if RFF helped or hurt.

**Action required IMMEDIATELY:**
1. Check the launch command actually used and report it here.
2. Check pod logs for any val-loop traceback.
3. Confirm whether sigma=10.0 was ever launched (and on which run id).
4. Decide whether to kill these and relaunch with validation working, or accept the lost compute and scope down.

Please respond within this polling cycle. We are burning GPU on a run we can't evaluate.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #88 because Closing as a dead end. After 25+ epochs of training, zero validation metrics were logged — we have no physically meaningful evaluation data to assess the RFF PE hypothesis. Additionally, the sigma sweep (sigma=1.0 vs sigma=10.0) that we agreed on never executed: all three runs logged sigma=1.0 in their configs. The experiment consumed a full GPU budget and produced no comparable results.

The RFF PE hypothesis itself remains scientifically valid — random Fourier features are a well-founded approach for approximating shift-invariant kernels on unstructured point clouds. If a follow-up is warranted, it should: (1) verify validation is working before committing to a 50-epoch run, (2) confirm sigma values in W&B config, (3) run a proper sigma sweep across at least two octaves (1.0 and 10.0)..
